### PR TITLE
Encoding of each type in JSON ABI and other enhancements

### PR DESCRIPTION
**Pull request by mohammadfawaz**
Closes https://github.com/FuelLabs/fuel-specs/issues/319

Also:
* Some general enhancements
* Improved and more illustrative examples
* Removed the type `Byte`.

### CODE REVIEW

File: specs/protocol/abi.md [hunk 1]
```diff
@@ -6,92 +6,333 @@ This document describes and specifies the ABI (Application Binary Interface) of
...
+    Foo: u64,
+    Bar: bool,
+}
+struct MyStruct {
+    bim: u8,
+    bam: MyEnum,
+}
+
+abi MyContract {
+    fn complex_function(arg1: MyStruct, arg2: [b256; 4], arg3: (str[5], bool)) -> str[6];
```
Comment (adlerjohn) [inline specs/protocol/abi.md:257]: Can you make this example be an array of structs? It's not actually clear how that would look like.

File: specs/protocol/abi.md [hunk 2]
```diff
@@ -137,11 +378,11 @@ _Important note:_ For the JSON representation of receipts, we represent 64-bit u
 
 ```json
 {
-    "type":"Panic",
-    "id":"0x39150017c9e38e5e280432d546fae345d6ce6d8fe4710162c2e3a95a6faff051",
-    "reason":"1",
-    "pc":"0xffffffffffffffff",
-    "is":"0xfffffffffffffffe"
+   "type":"Panic",
```
Comment (adlerjohn) [inline specs/protocol/abi.md:381]: Why go from 4 to 3 indentation?

File: specs/protocol/abi.md (hunk 1 above)
Comment (mohammadfawaz) [reply specs/protocol/abi.md]: Done.

File: specs/protocol/abi.md (hunk 2 above)
Comment (mohammadfawaz) [reply specs/protocol/abi.md]: So at first I thought `forc run --pretty-print` uses 3 spaces so I was running with 3 everywhere. Turns out that whatever JSON library we're using actually uses 2 spaces. Same goes for the JSON ABI file that the compiler generate. So, I'm now using 2 spaces everywhere.

**Review by adlerjohn — APPROVED** [review]

**Pull request merged**